Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m

```diff
@@ -0,0 +1,356 @@
+    % else
+    %     x.Q.InitialPLD_PWI4D = x.Q.Initial_PLD;
+    % end
+    % nPLD = length(x.Q.InitialPLD_PWI4D);
+    % 
+
+
+    
+
+    %% 3) Labeling duration (LD) vectors (identical to below)
+    % First we deal with a too long vector (e.g., cutting off control volumes for Hadamard)
+    nLabDur = length(x.Q.LabelingDuration);
+    if nLabDur>nVolumes
+        x.Q.LabelingDuration_PWI4D = x.Q.LabelingDuration(1:nVolumes); % can we do this based on the ASL4Dcontext.tsv ?
+    else
+        x.Q.LabelingDuration_PWI4D = x.Q.LabelingDuration;
+    end
+    nLabDur = length(x.Q.LabelingDuration_PWI4D);
+
+
+    end
+
+
+    %% TIME DECODING WARNING -> MOVE TO xASL_quant_HadamardDecoding (check the original code)
+    %%  BECAUSE TIME DECODING WILL ALSO ADAPT THE QUANTIFICATION PARAMETERS
+    %%  ORDER
+    % %% This part needs to be synced with the below part still
+    % %% That we give equal warnings for all ASL flavor, if the number of TE/PLD/labdurs 
+    % %% don't fit within the number of volumes
+    % %% unless we already give this in the import, then we can remove it here
+    % 
+    % % nVolumesPerRepetition = nTEs .* (nUnique_InitialPLD - (nUnique_InitialPLD./x.Q.TimeEncodedMatrixSize));
+    % nVolumesPerRepetition = nTEs .* nUnique_InitialPLD;
+    % 
+    % % nVolumesPerRepetition is also called blockSize by some
+	% nRepetitions = nVolumes ./ nVolumesPerRepetition;
+    % 
+    % % First check if the number of volumes fits with the number of
+    % % expected volumes per repetition (for the amount of x.Q.EchoTimes, PLDs, and labeling durations)
+    % if nRepetitions~=round(nRepetitions)
+    %     error(['nVolumes ' xASL_num2str(size(PWI4D,4)) ' cannot be composed of ' xASL_num2str(nVolumesPerRepetition) ' volumes per repetition']);
+    % end
+
+
+
+    
+    
+% =====================================================================
+% C) Single- and multi-PLD subtraction
+elseif bCreatePWI4D
+    %% 4. Create PWI4D
+    if nVolumes>1 && ~x.modules.asl.bContainsDeltaM
+        % Paired subtraction
+        [ControlIm, LabelIm] = xASL_quant_GetControlLabelOrder(ASL4D);
+        PWI4D = ControlIm - LabelIm;
+        
+        % Skip every other value in the vectors as they were stored for both control and label images 
+        x.Q.EchoTime_PWI4D = x.Q.EchoTime(1:2:end);
+        x.Q.InitialPLD_PWI4D = x.Q.Initial_PLD(1:2:end);
+        x.Q.LabelingDuration_PWI4D = x.Q.LabelingDuration(1:2:end);
+    else % the same but then without subtraction
+        PWI4D = ASL4D;
+    end
+
+end
+   
+if bCreatePWI3D
+    %% 5. Create PWI3D
+    
+    % After averaging across PLDs, we'll obtain these unique PLDs+LD combinations
+    % indexAverage_PLD_LabDur lists for each original position to where it should be averaged
+    [~, ~, iUnique_TE_PLD_LabDur] = unique([x.Q.EchoTime_PWI4D(:), x.Q.InitialPLD_PWI4D(:), x.Q.LabelingDuration_PWI4D(:)], 'stable', 'rows');
```

Comment:
Nice to merge this for hadamard and normal multi-pld. 
And this will nicely work. 

Just to show that there are parts I'm happy with ;)